shared/js/passcode_helper.js

```diff
@@ -0,0 +1,161 @@
+/* exported PasscodeHelper */
+/* globals crypto, TextEncoder */
+(function(exports) {
+  'use strict';
+
+  const SET_DIGEST_VALUE = 'lockscreen.passcode-lock.digest.value';
+  const SET_DIGEST_SALT = 'lockscreen.passcode-lock.digest.salt';
+  const SET_DIGEST_ITERATIONS = 'lockscreen.passcode-lock.digest.iterations';
+  const SET_DIGEST_ALGORITHM = 'lockscreen.passcode-lock.digest.algorithm';
+  const DEFAULT_ALGORITHM = 'SHA-1'; //XXX Update to SHA-256 after bug 554827?
+  const DEFAULT_ITERATIONS = 1000;
+
+
+  var PasscodeHelper = function() {
+    /* Usage:
+     *      var ph = new PasscodeHelper();
+     *      ph.setPassccode(string) -> Promise => digest (or false)
+     *      ph.checkPasscode(string) -> Promise resolves to a boolean
+     * */
+
+
+    var _salt;
+    var _iterations;
+    var _algorithm;
+    var _testPass;
+
+    this._encode = function(str) {
```

Comment:
Well, to prevent method copying you should use

`PasscodeHelper.prototype._encode`

instead of create and register methods in the constructor. The difference is if you create them in the constructor, it would be copied every times you create a new instance from this constructor, rather than refer to the same method already registered as a prototype object's method, which can save lots of memory.

Another minor reason is if this component keep growing, the constructor would be too long, and the notorious 'ending parenthesis'  issue appears.